diff -up dovecot-2.2.10/dovecot.service.in.waitonline dovecot-2.2.10/dovecot.service.in
--- dovecot-2.2.10/dovecot.service.in.waitonline	2016-06-09 17:07:30.965683169 +0200
+++ dovecot-2.2.10/dovecot.service.in	2016-06-09 17:08:56.257389557 +0200
@@ -1,9 +1,10 @@
 [Unit]
 Description=Dovecot IMAP/POP3 email server
-After=local-fs.target network.target
+After=local-fs.target network.target network-online.target
 
 [Service]
 Type=simple
+ExecStartPre=/usr/libexec/dovecot/prestartscript
 ExecStart=@sbindir@/dovecot -F
 ExecReload=/bin/kill -HUP $MAINPID
 PrivateTmp=true
